File: skills/sci-plot/assets/case-packs/rf-0054/gene_cluster_enrichment_heatmap.R
args <- commandArgs(trailingOnly = FALSE)
fa <- grep("^--file=", args, value = TRUE)
base_dir <- if (length(fa)) dirname(normalizePath(sub("^--file=", "", fa[1]), winslash = "/", mustWork = FALSE)) else getwd()
cluster_cols <- c("#1F77B4", "#FF7F0E", "#2CA02C", "#D62728", "#9467BD", "#8C564B", "#E377C2", "#7F7F7F", "#BCBD22")
heat_cols <- colorRampPalette(c("#164E8A", "#F7F5F2", "#B3202A"))(201)
asg <- read.csv(file.path(base_dir, "gene_cluster_assignments.csv"), stringsAsFactors = FALSE)
enr <- read.csv(file.path(base_dir, "enrichment_annotations.csv"), stringsAsFactors = FALSE)
zrow <- function(x) {
  m <- rowMeans(x)
  s <- apply(x, 1, sd)
  s[!is.finite(s) | s == 0] <- 1
  (x - m) / s
}
stable_order <- function(x) {
  if (nrow(x) < 3) return(seq_len(nrow(x)))
  y <- sweep(x, 2, colMeans(x))
  v <- svd(y, nu = 0, nv = 1)$v[, 1]
  f <- which(abs(v) > 1e-12)[1]
  if (length(f) && v[f] < 0) v <- -v
  order(as.vector(y %*% v), method = "radix")
}
draw_profile <- function(values, blocks, samples) {
  n <- nrow(values)
  plot.new(); plot.window(c(-.38, 1.05), c(n, 0), xaxs = "i", yaxs = "i")
  title("Cluster profile", line = .6, cex.main = .83, font.main = 2, col.main = "#26323D")
  xx <- seq(.04, .97, length.out = ncol(values))
  for (k in seq_len(nrow(blocks))) {
    b <- blocks[k, ]; cl <- b$cluster; st <- b$start; en <- b$end; h <- en - st
    rect(-.06, st, 1.02, en, col = if (k %% 2) "#F6F7F8" else "white", border = "#D8DDE2", lwd = .7)
    med <- apply(values[(st + 1):en, , drop = FALSE], 2, median)
    amp <- max(abs(med), .5)
    yy <- (st + en) / 2 - med / amp * h * .28
    lines(xx, yy, col = "white", lwd = 4)
    lines(xx, yy, col = cluster_cols[cl], lwd = 2.2)
    text(-.34, (st + en) / 2, paste0("C", cl), adj = 0, font = 2, cex = .76, col = "#26323D")
    text(-.02, st + h * .12, paste0("n = ", h), adj = c(0, 1), cex = .48, col = "#5E6A74")
  }
}
draw_genes <- function(genes, blocks, reps) {
  n <- length(genes)
  plot.new(); plot.window(c(0, 1), c(n, 0), xaxs = "i", yaxs = "i")
  title("Representative genes", line = .6, cex.main = .83, font.main = 2, col.main = "#26323D")
  for (k in seq_len(nrow(blocks))) {
    b <- blocks[k, ]; wanted <- reps[[as.character(b$cluster)]]
    wanted <- wanted[wanted %in% genes[(b$start + 1):b$end]]
    if (!length(wanted)) next
    targets <- seq(b$start + (b$end-b$start)*.16, b$end-(b$end-b$start)*.16, length.out = length(wanted))
    actual <- match(wanted, genes) - .5
    for (j in seq_along(wanted)) {
      segments(.87, targets[j], 1, actual[j], col = "#8B949C", lwd = .65)
      text(.83, targets[j], wanted[j], adj = 1, cex = .53, font = 3, col = "#26323D")
    }
  }
}
draw_heat <- function(values, samples, sample_cols, blocks) {
  n <- nrow(values); p <- ncol(values)
  plot.new(); plot.window(c(.5, p + .5), c(n, 0), xaxs = "i", yaxs = "i")
  z <- values
  z[z < -2] <- -2
  z[z > 2] <- 2
  image(x = seq_len(p), y = seq(.5, n - .5), z = t(z[n:1, , drop = FALSE]), col = heat_cols, zlim = c(-2, 2), add = TRUE, useRaster = TRUE)
  for (j in seq_len(p)) rect(j-.5, -n*.024, j+.5, -n*.008, col = sample_cols[j], border = "white", xpd = NA)
  for (j in seq_len(p)) text(j, -n*.035, samples[j], srt = 40, adj = 0, cex = .54, xpd = NA, col = "#34414A")
  for (y in blocks$end) segments(.5, y, p+.5, y, col = "#D4D9DD", lwd = .6)
  box(col = "#26323D", lwd = .8)
}
draw_stripe <- function(blocks, n) {
  plot.new(); plot.window(c(0,1), c(n,0), xaxs="i", yaxs="i")
  for (k in seq_len(nrow(blocks))) {
    b <- blocks[k, ]; rect(0,b$start,1,b$end,col=cluster_cols[b$cluster],border="white")
    text(.5,(b$start+b$end)/2,paste0("C",b$cluster,"  n=",b$end-b$start),srt=90,col="white",font=2,cex=.42)
  }
}
draw_terms <- function(blocks, enrichment, n, dataset) {
  plot.new(); plot.window(c(0,1), c(n,0), xaxs="i", yaxs="i")
  title("Functional enrichment", line=.6, cex.main=.83, font.main=2, col.main="#26323D")
  for (k in seq_len(nrow(blocks))) {
    b <- blocks[k, ]; rect(0,b$start,1,b$end,col=if(k%%2)"#F7F7F6" else "#FBFBFA",border="#AEB6BC",lwd=.65)
    q <- enrichment[enrichment$cluster == b$cluster, ]
    q <- q[order(q$rank), ]
    if (dataset == "pbmc" && b$end-b$start <= 5) q <- head(q,3)
    yy <- seq(b$start+(b$end-b$start)*.14,b$end-(b$end-b$start)*.14,length.out=nrow(q))
    for (j in seq_len(nrow(q))) text(.018,yy[j],q$term[j],adj=0,cex=if(dataset=="pbmc").47 else .51,font=if(j==1)2 else 1,col=cluster_cols[b$cluster])
  }
}
draw_bars <- function(blocks, enrichment, n, dataset) {
  plot.new(); plot.window(c(0,1.04), c(n,0), xaxs="i", yaxs="i")
  title("Relative enrichment", line=.6, cex.main=.83, font.main=2, col.main="#26323D")
  for (k in seq_len(nrow(blocks))) {
    b <- blocks[k, ]; q <- enrichment[enrichment$cluster == b$cluster, ]; q <- q[order(q$rank), ]
    if (dataset == "pbmc" && b$end-b$start <= 5) q <- head(q,3)
    yy <- seq(b$start+(b$end-b$start)*.14,b$end-(b$end-b$start)*.14,length.out=nrow(q))
    hh <- max(n*.0015,(b$end-b$start)/(max(1,nrow(q))*2.8))
    for (j in seq_len(nrow(q))) { rect(0,yy[j]-hh/2,1,yy[j]+hh/2,col="#EEF0F2",border=NA); rect(0,yy[j]-hh/2,q$relative_score[j],yy[j]+hh/2,col=cluster_cols[b$cluster],border=NA) }
  }
}
render <- function(dataset, genes, samples, values, clusters, scores, cluster_order, sample_cols, reps, title_text, subtitle_text, output) {
  ord <- integer(); starts <- numeric(); ends <- numeric()
  for (cl in cluster_order) {
    ii <- which(clusters == cl)
    if (dataset == "embryonic") ii <- ii[stable_order(values[ii, , drop=FALSE])]
    starts <- c(starts,length(ord)); ord <- c(ord,ii); ends <- c(ends,length(ord))
  }
  values <- values[ord,,drop=FALSE]; genes <- genes[ord]; clusters <- clusters[ord]
  blocks <- data.frame(cluster=cluster_order,start=starts,end=ends)
  enrichment <- enr[enr$dataset == dataset,]
  png(file.path(base_dir,output),width=4400,height=2500,res=220,bg="white")
  par(oma=c(3.1,3.2,7.5,2),mar=c(.5,.2,2.6,.2),family="sans")
  layout(matrix(1:6,1,6),widths=c(1.7,1.55,1.6,.22,5.6,1.55))
  draw_profile(values,blocks,samples)
  draw_genes(genes,blocks,reps)
  draw_heat(values,samples,sample_cols,blocks)
  draw_stripe(blocks,nrow(values))
  draw_terms(blocks,enrichment,nrow(values),dataset)
  draw_bars(blocks,enrichment,nrow(values),dataset)
  mtext(title_text,outer=TRUE,side=3,line=4.6,adj=0,font=2,cex=1.7,col="#202A33")
  mtext(subtitle_text,outer=TRUE,side=3,line=2.8,adj=0,cex=.9,col="#687680")
  mtext("Rows are genes; colors show row-wise z-scores. Profiles summarize median expression within each cluster.",outer=TRUE,side=1,line=1.2,adj=0,cex=.72,col="#687680")
  dev.off()
}
emb <- read.csv(file.path(base_dir,"embryonic_expression.csv"),check.names=FALSE)
ea <- asg[asg$dataset=="embryonic",]
ix <- match(emb$gene,ea$gene)
stopifnot(nrow(emb)==3767,all(!is.na(ix)))
ereps <- list(`1`=c("Apeh","Ckb","Nit2"),`2`=c("Abcf1","Top2a","Ipo5"),`3`=c("Fam192a","Ndufb11"),`4`=c("Fem1b","Thrap3","Peci","Psmd11"),`5`=c("Tbc1d23","Dusp7","Cdk5rap2"),`6`=c("Rab27a","Dtx2","Acad10","AU015836"),`7`=c("Morc3","Yipf3","Luc7l2","Timm8a1"),`8`=c("Tbcc","Atp11c","Tm7sf2","LOC100503972"))
render("embryonic",emb$gene,c("Zygote","2-cell","4-cell","8-cell","Morula","Blastocyst"),zrow(as.matrix(emb[,-1])),ea$cluster[ix],ea$score[ix],c(6,1,8,5,3,7,4,2),c("#F1D94F","#7776DE","#D67812","#75DCA2","#F27418","#1768B5"),ereps,"Embryonic expression programs","Fuzzy clustering of 3,767 genes across six stages of preimplantation development","embryonic_gene_cluster_heatmap_r.png")
pb <- read.csv(file.path(base_dir,"pbmc_normalized_expression.csv"),check.names=FALSE)
meta <- read.csv(file.path(base_dir,"pbmc_cell_metadata.csv"),stringsAsFactors=FALSE)
pa <- asg[asg$dataset=="pbmc",]
levels_pb <- c("Naive CD4 T","Memory CD4 T","CD14+ Mono","B","CD8 T","FCGR3A+ Mono","NK","DC","Platelet")
cell_type <- meta$cell_type[match(names(pb)[-1],meta$cell)]
lin <- expm1(as.matrix(pb[,-1]))
avg <- sapply(levels_pb,function(z) rowMeans(lin[,cell_type==z,drop=FALSE]))
sel <- match(pa$gene,pb$gene)
preps <- list(`1`=c("RPS23","FUS","RPSA"),`2`=c("GIMAP7","CRIP1","PRDX2"),`3`=c("GPX1","JUND","LYL1","VAMP5"),`4`=c("PKIG","POLD4","SELL"),`5`=c("SH2D1A","BIN2","CELF2"),`6`=c("MBD2","CXCL16","TMEM127","RAB31"),`7`=c("C19orf10","MAD2L2","TMBIM4.1","CARD8"),`8`=c("TRAPPC9","AMPD2","ENHO","PRR14L"),`9`=c("CD9","RBBP6","GNG11"))
render("pbmc",pa$gene,levels_pb,zrow(avg[sel,,drop=FALSE]),pa$cluster,pa$score,1:9,c("#7A9A01","#4DE88D","#62E679","#3A12C6","#84AAA0","#D000B9","#168CA8","#71DEDC","#58EFB5"),preps,"Cell-type marker programs","Differential expression and functional enrichment across nine PBMC lineages","pbmc_gene_cluster_heatmap_r.png")

File: skills/sci-plot/assets/case-packs/rf-0054/gene_cluster_enrichment_heatmap.py
from pathlib import Path
import csv
import numpy as np
import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.colors import LinearSegmentedColormap
from matplotlib.patches import Rectangle


BASE = Path(__file__).resolve().parent
CLUSTER_COLORS = {1: "#1f77b4", 2: "#ff7f0e", 3: "#2ca02c", 4: "#d62728", 5: "#9467bd", 6: "#8c564b", 7: "#e377c2", 8: "#7f7f7f", 9: "#bcbd22"}
HEATMAP = LinearSegmentedColormap.from_list("expression", ["#164e8a", "#f7f5f2", "#b3202a"])


def read_matrix(filename):
    with (BASE / filename).open(newline="", encoding="utf-8-sig") as handle:
        reader = csv.reader(handle)
        header = next(reader)
        rows = list(reader)
    genes = [row[0] for row in rows]
    values = np.array([[float(value) for value in row[1:]] for row in rows], dtype=float)
    if len(genes) != len(set(genes)) or not np.isfinite(values).all():
        raise ValueError(f"Invalid matrix in {filename}")
    return genes, header[1:], values


def read_assignments():
    result = {"embryonic": [], "pbmc": []}
    with (BASE / "gene_cluster_assignments.csv").open(newline="", encoding="utf-8-sig") as handle:
        for row in csv.DictReader(handle):
            result[row["dataset"]].append((row["gene"], int(row["cluster"]), float(row["score"])))
    return result


def read_enrichment():
    result = {"embryonic": {}, "pbmc": {}}
    with (BASE / "enrichment_annotations.csv").open(newline="", encoding="utf-8-sig") as handle:
        for row in csv.DictReader(handle):
            dataset = row["dataset"]
            cluster = int(row["cluster"])
            result[dataset].setdefault(cluster, []).append((int(row["rank"]), row["term"], float(row["relative_score"])))
    for dataset in result:
        for cluster in result[dataset]:
            result[dataset][cluster].sort(key=lambda item: item[0])
    return result


def row_zscore(values):
    centered = values - values.mean(axis=1, keepdims=True)
    spread = values.std(axis=1, ddof=1, keepdims=True)
    spread[spread == 0] = 1
    return centered / spread


def stable_profile_order(values):
    if len(values) < 3:
        return np.arange(len(values))
    centered = values - values.mean(axis=0, keepdims=True)
    _, _, vectors = np.linalg.svd(centered, full_matrices=False)
    direction = vectors[0]
    first = np.flatnonzero(np.abs(direction) > 1e-12)
    if len(first) and direction[first[0]] < 0:
        direction = -direction
    projection = np.sum(centered * direction[None, :], axis=1)
    return np.argsort(projection, kind="stable")


def prepare_embryonic(assignments):
    genes, samples, expression = read_matrix("embryonic_expression.csv")
    if expression.shape != (3767, 6):
        raise ValueError("Embryonic expression matrix must contain 3,767 genes and 6 stages")
    assignment = {gene: (cluster, score) for gene, cluster, score in assignments}
    if set(assignment) != set(genes):
        raise ValueError("Embryonic cluster assignments do not match the expression matrix")
    clusters = np.array([assignment[gene][0] for gene in genes], dtype=int)
    scores = np.array([assignment[gene][1] for gene in genes], dtype=float)
    expected = {1: 372, 2: 281, 3: 282, 4: 847, 5: 250, 6: 424, 7: 961, 8: 350}
    observed = {cluster: int((clusters == cluster).sum()) for cluster in expected}
    if observed != expected:
        raise ValueError(f"Unexpected embryonic cluster sizes: {observed}")
    labels = ["Zygote", "2-cell", "4-cell", "8-cell", "Morula", "Blastocyst"]
    colors = ["#f1d94f", "#7776de", "#d67812", "#75dca2", "#f27418", "#1768b5"]
    representatives = {1: ["Apeh", "Ckb", "Nit2"], 2: ["Abcf1", "Top2a", "Ipo5"], 3: ["Fam192a", "Ndufb11"], 4: ["Fem1b", "Thrap3", "Peci", "Psmd11"], 5: ["Tbc1d23", "Dusp7", "Cdk5rap2"], 6: ["Rab27a", "Dtx2", "Acad10", "AU015836"], 7: ["Morc3", "Yipf3", "Luc7l2", "Timm8a1"], 8: ["Tbcc", "Atp11c", "Tm7sf2", "LOC100503972"]}
    return genes, labels, row_zscore(expression), clusters, scores, [6, 1, 8, 5, 3, 7, 4, 2], colors, representatives


def prepare_pbmc(assignments):
    genes, cells, expression = read_matrix("pbmc_normalized_expression.csv")
    if expression.shape != (500, 500):
        raise ValueError("PBMC expression matrix must contain 500 genes and 500 cells")
    with (BASE / "pbmc_cell_metadata.csv").open(newline="", encoding="utf-8-sig") as handle:
        metadata = {row["cell"]: row["cell_type"] for row in csv.DictReader(handle)}
    levels = ["Naive CD4 T", "Memory CD4 T", "CD14+ Mono", "B", "CD8 T", "FCGR3A+ Mono", "NK", "DC", "Platelet"]
    labels = np.array([metadata[cell] for cell in cells])
    if set(labels) != set(levels):
        raise ValueError("PBMC cell annotations are incomplete")
    marker_genes = [item[0] for item in assignments]
    index = {gene: position for position, gene in enumerate(genes)}
    if len(marker_genes) != 77 or any(gene not in index for gene in marker_genes):
        raise ValueError("PBMC marker assignments are incomplete")
    linear = np.expm1(expression)
    average = np.column_stack([linear[:, labels == level].mean(axis=1) for level in levels])
    selected = average[[index[gene] for gene in marker_genes]]
    clusters = np.array([item[1] for item in assignments], dtype=int)
    scores = np.array([item[2] for item in assignments], dtype=float)
    expected = [6, 6, 15, 5, 4, 12, 6, 18, 5]
    observed = [int((clusters == cluster).sum()) for cluster in range(1, 10)]
    if observed != expected:
        raise ValueError(f"Unexpected PBMC marker cluster sizes: {observed}")
    colors = ["#7a9a01", "#4de88d", "#62e679", "#3a12c6", "#84aaa0", "#d000b9", "#168ca8", "#71dedc", "#58efb5"]
    representatives = {1: ["RPS23", "FUS", "RPSA"], 2: ["GIMAP7", "CRIP1", "PRDX2"], 3: ["GPX1", "JUND", "LYL1", "VAMP5"], 4: ["PKIG", "POLD4", "SELL"], 5: ["SH2D1A", "BIN2", "CELF2"], 6: ["MBD2", "CXCL16", "TMEM127", "RAB31"], 7: ["C19orf10", "MAD2L2", "TMBIM4.1", "CARD8"], 8: ["TRAPPC9", "AMPD2", "ENHO", "PRR14L"], 9: ["CD9", "RBBP6", "GNG11"]}
    return marker_genes, levels, row_zscore(selected), clusters, scores, list(range(1, 10)), colors, representatives


def render_figure(dataset, genes, samples, values, clusters, scores, cluster_order, sample_colors, representatives, enrichment, title, subtitle, output):
    order = []
    blocks = []
    start = 0
    for cluster in cluster_order:
        positions = np.where(clusters == cluster)[0]
        local = stable_profile_order(values[positions]) if dataset == "embryonic" else np.arange(len(positions))
        positions = positions[local]
        order.extend(positions.tolist())
        end = start + len(positions)
        blocks.append((cluster, start, end, len(positions)))
        start = end
    order = np.array(order, dtype=int)
    displayed_values = values[order]
    displayed_genes = [genes[index] for index in order]
    displayed_clusters = clusters[order]
    displayed_scores = scores[order]
    total = len(displayed_genes)
    figure = plt.figure(figsize=(20, 12), facecolor="white")
    grid = figure.add_gridspec(1, 6, width_ratios=[1.7, 1.55, 1.6, 0.22, 5.6, 1.55], left=0.045, right=0.965, bottom=0.075, top=0.84, wspace=0.035)
    profile_axis = figure.add_subplot(grid[0, 0])
    gene_axis = figure.add_subplot(grid[0, 1], sharey=profile_axis)
    heat_axis = figure.add_subplot(grid[0, 2], sharey=profile_axis)
    stripe_axis = figure.add_subplot(grid[0, 3], sharey=profile_axis)
    term_axis = figure.add_subplot(grid[0, 4], sharey=profile_axis)
    bar_axis = figure.add_subplot(grid[0, 5], sharey=profile_axis)
    image = heat_axis.imshow(displayed_values, aspect="auto", interpolation="nearest", cmap=HEATMAP, vmin=-2, vmax=2, extent=(-0.5, len(samples) - 0.5, total, 0), rasterized=True)
    heat_axis.set_xlim(-0.5, len(samples) - 0.5)
    heat_axis.set_ylim(total, 0)
    heat_axis.set_yticks([])
    heat_axis.set_xticks(np.arange(len(samples)))
    heat_axis.set_xticklabels(samples, rotation=40, ha="left", va="bottom", fontsize=9.2)
    heat_axis.tick_params(axis="x", top=True, labeltop=True, bottom=False, labelbottom=False, length=0, pad=20)
    for position, color in enumerate(sample_colors):
        heat_axis.add_patch(Rectangle((position - 0.5, -total * 0.024), 1, total * 0.016, transform=heat_axis.transData, facecolor=color, edgecolor="white", linewidth=0.7, clip_on=False))
    for spine in heat_axis.spines.values():
        spine.set_color("#26323d")
        spine.set_linewidth(0.9)
    for axis in [profile_axis, gene_axis, stripe_axis, term_axis, bar_axis]:
        axis.set_ylim(total, 0)
        axis.set_yticks([])
    profile_axis.set_xlim(-0.38, 1.04)
    gene_axis.set_xlim(0, 1)
    stripe_axis.set_xlim(0, 1)
    term_axis.set_xlim(0, 1)
    bar_axis.set_xlim(0, 1.04)
    for axis in [profile_axis, gene_axis, stripe_axis, term_axis, bar_axis]:
        axis.set_xticks([])
        for spine in axis.spines.values():
            spine.set_visible(False)
    profile_axis.text(0.38, 1.035, "Cluster profile", transform=profile_axis.transAxes, ha="center", va="bottom", fontsize=10.5, fontweight="bold", color="#26323d")
    gene_axis.text(0.95, 1.035, "Representative genes", transform=gene_axis.transAxes, ha="right", va="bottom", fontsize=10.5, fontweight="bold", color="#26323d")
    term_axis.text(0.02, 1.035, "Functional enrichment", transform=term_axis.transAxes, ha="left", va="bottom", fontsize=10.5, fontweight="bold", color="#26323d")
    bar_axis.text(0.5, 1.035, "Relative enrichment", transform=bar_axis.transAxes, ha="center", va="bottom", fontsize=10.5, fontweight="bold", color="#26323d")
    gene_position = {gene: position + 0.5 for position, gene in enumerate(displayed_genes)}
    sample_x = np.linspace(0.02, 0.98, len(samples))
    for block_number, (cluster, block_start, block_end, count) in enumerate(blocks):
        height = block_end - block_start
        center = (block_start + block_end) / 2
        color = CLUSTER_COLORS[cluster]
        profile_axis.add_patch(Rectangle((-0.06, block_start), 1.08, height, facecolor="#f6f7f8" if block_number % 2 == 0 else "#ffffff", edgecolor="#d8dde2", linewidth=0.8))
        block_values = displayed_values[displayed_clusters == cluster]
        median = np.median(block_values, axis=0)
        amplitude = max(float(np.max(np.abs(median))), 0.5)
        profile_y = center - median / amplitude * height * 0.28
        profile_axis.plot(sample_x, profile_y, color="white", linewidth=4.2, solid_capstyle="round", zorder=2)
        profile_axis.plot(sample_x, profile_y, color=color, linewidth=2.4, solid_capstyle="round", zorder=3)
        profile_axis.text(-0.34, center, f"C{cluster}", ha="left", va="center", fontsize=11.5, fontweight="bold", color="#26323d")
        profile_axis.text(-0.02, block_start + height * 0.12, f"n = {count}", ha="left", va="top", fontsize=7.8, color="#5e6a74")
        candidates = [gene for gene in representatives.get(cluster, []) if gene in gene_position and displayed_clusters[int(gene_position[gene] - 0.5)] == cluster]
        if candidates:
            label_y = np.linspace(block_start + height * 0.16, block_end - height * 0.16, len(candidates))
            for gene, target_y in zip(candidates, label_y):
                actual_y = gene_position[gene]
                gene_axis.plot([0.88, 1], [target_y, actual_y], color="#8b949c", linewidth=0.75, clip_on=False)
                gene_axis.text(0.84, target_y, gene, ha="right", va="center", fontsize=7.9 if dataset == "embryonic" else 8.6, fontstyle="italic", color="#26323d")
        stripe_axis.add_patch(Rectangle((0, block_start), 1, height, facecolor=color, edgecolor="white", linewidth=1.1))
        stripe_axis.text(0.5, center, f"C{cluster}  n={count}", rotation=90, ha="center", va="center", fontsize=7.2, fontweight="bold", color="white")
        term_axis.add_patch(Rectangle((0, block_start), 1, height, facecolor="#f7f7f6" if block_number % 2 == 0 else "#fbfbfa", edgecolor="#aeb6bc", linewidth=0.8))
        terms = enrichment[cluster]
        if dataset == "pbmc" and height <= 5:
            terms = terms[:3]
        term_y = np.linspace(block_start + height * 0.14, block_end - height * 0.14, len(terms))
        font_size = 8.0 if dataset == "pbmc" else 8.7
        for term_index, ((rank, term, relative_score), y_value) in enumerate(zip(terms, term_y)):
            term_axis.text(0.025, y_value, term, ha="left", va="center", fontsize=font_size + (0.8 if term_index == 0 else 0), fontweight="bold" if term_index == 0 else "normal", color=color, clip_on=True)
            bar_height = max(total * 0.0015, height / (len(terms) * 2.8))
            bar_axis.barh(y_value, 1, height=bar_height, color="#eef0f2", edgecolor="none")
            bar_axis.barh(y_value, relative_score, height=bar_height, color=color, edgecolor="none")
        for axis in [gene_axis, heat_axis, stripe_axis, bar_axis]:
            axis.axhline(block_end, color="#d4d9dd", linewidth=0.65, zorder=4)
    figure.text(0.045, 0.958, title, ha="left", va="top", fontsize=23, fontweight="bold", color="#202a33")
    figure.text(0.045, 0.922, subtitle, ha="left", va="top", fontsize=11.5, color="#687680")
    figure.text(0.045, 0.032, "Rows are genes; colors show row-wise z-scores. Profiles summarize median expression within each cluster.", ha="left", va="center", fontsize=8.8, color="#687680")
    color_axis = figure.add_axes([0.82, 0.026, 0.13, 0.014])
    colorbar = figure.colorbar(image, cax=color_axis, orientation="horizontal", ticks=[-2, 0, 2])
    colorbar.ax.tick_params(labelsize=7.8, length=2, colors="#49545d", pad=1)
    colorbar.outline.set_visible(False)
    colorbar.ax.set_title("Row z-score", fontsize=8.2, color="#49545d", pad=3)
    figure.savefig(BASE / output, dpi=220, facecolor="white", edgecolor="none")
    plt.close(figure)


def main():
    plt.rcParams.update({"font.family": "sans-serif", "font.sans-serif": ["Arial", "Helvetica", "DejaVu Sans"], "axes.unicode_minus": False, "savefig.facecolor": "white"})
    assignments = read_assignments()
    enrichment = read_enrichment()
    embryo = prepare_embryonic(assignments["embryonic"])
    render_figure("embryonic", *embryo, enrichment["embryonic"], "Embryonic expression programs", "Fuzzy clustering of 3,767 genes across six stages of preimplantation development", "embryonic_gene_cluster_heatmap.png")
    pbmc = prepare_pbmc(assignments["pbmc"])
    render_figure("pbmc", *pbmc, enrichment["pbmc"], "Cell-type marker programs", "Differential expression and functional enrichment across nine PBMC lineages", "pbmc_gene_cluster_heatmap.png")


if __name__ == "__main__":
    main()
